Include one_hot_active_task in MultiEnvWrapper

nicolengsy opened the pull request:
To resolve issue #1868

ryanjulian commented on src/garage/envs/multi_env_wrapper.py:
```diff
@@ -187,7 +187,7 @@ def reset(self):
             pass
         elif self._mode == 'add-onehot':
-            obs = np.concatenate([obs, self._active_task_one_hot()])
+            episode_info["active_task_one_hot"] = self._active_task_one_hot()
```
did you install pre-commit? we only use single quotes

ryanjulian replied:
```suggestion
            episode_info['task_one_hot'] = self._active_task_one_hot()
```

irisliucy commented:
@nicolengsy Good job! Look like @ryanjulian was right, some tests in `test_mtsac.py` are broken. In particular, I think you need to change `torch.Tensor([env.reset()[0]]* buffer_batch_size)` to `torch.Tensor([np.concatenate([env.reset()[0], env.reset()[1]['task_one_hot']])] * buffer_batch_size)` under `test_mtsac_get_log_alpha()` and maybe some other things too.

nicolengsy commented:
Thanks, made the requested changes! Didn't include it in this PR, but moving forward wondering if we'd want to get rid of the `mode` argument when initializing MultiEnvWrapper, since it was previously used to control whether one_hot info was included in observations, which won't happen anymore.

yeukfu commented on src/garage/envs/multi_env_wrapper.py:
```diff
@@ -106,17 +106,7 @@ def observation_space(self):
-            env_lb, env_ub = self._env.observation_space.bounds
-            return akro.Box(np.concatenate([env_lb, task_lb]),
-                            np.concatenate([env_ub, task_ub]))
-        else:  # self._mode == 'del-onehot'
```
If 'del-onehot' mode is kept. I think we should still process it.

yeukfu commented on src/garage/envs/multi_env_wrapper.py:
```diff
@@ -204,13 +194,7 @@ def step(self, action):
-
-        if self._mode == 'add-onehot':
-            obs = np.concatenate([es.observation, self._active_task_one_hot()])
-        elif self._mode == 'del-onehot':
```
If 'del-onehot' mode is kept. We still need to remove these from obs?

AiRuiChen replied:
Is mode `del-onehot` ever used? IIUC only MultiEnvWrapper will populate the active task index in observation. After this PR, the index goes into `episode_info`. So I believe there is no need to remove task index from `obs` (aka mode `del-onehot`), and we can simply do `obs = es.observation` here

yeukfu reviewed (approved):
Looks good to me. I think you should also update the 'mode' docstring.

krzentner commented:
Wait, I don't understand this change at all. Why are we just removing onehot support from `MultiEnvWrapper`? That doesn't look like what the original bug report was describing.

nicolengsy commented:
@krzentner I've removed one-hot from the `observations`, and moved it into `episode_info` instead (see line 180 in `multi_env_wrapper.py`). I deleted checks for the mode (i.e. `add-onehot`), since one-hot is always included in `episode_info` now. I should refactor the mode out of all the files to be consistent, which I can do within this PR as well.

Was this not what Eric asked for in #1868 ?

AiRuiChen commented on src/garage/envs/multi_env_wrapper.py:
```diff
@@ -187,7 +177,7 @@ def reset(self):
-            obs = np.concatenate([obs, self._active_task_one_hot()])
+            episode_info['task_one_hot'] = self._active_task_one_hot()
         else:  # self._mode == 'del-onehot'
             obs = obs[:-self._num_tasks]
```
I'd suggest to remove `mode` from MultiEnvWrapper in this PR. The active task index should *always* be in `episode_info`, we don't need a mode to determine that. 
Once you done that you can remove the if conditions here and simply do `episode_info['task_one_hot'] = self._active_task_one_hot()`.

AiRuiChen commented:
> Wait, I don't understand this change at all. Why are we just removing onehot support from `MultiEnvWrapper`? That doesn't look like what the original bug report was describing.

@krzentner The issue description was wrong and I just updated it. I meant to put `MultiEnvWrapper`'s active task index in `episode_info`, instead of concatenating it to `obs`. 

This PR does the described functionality. But I think what's implied in that issue is: we should also make all downstream algos to consume active task index from `episode_info`, not from`obs`. @nicolengsy This future change can go into another PR. It's fine now to manually plug active task index back to `obs` before feeding into algos (like what you did in test_mtsac.py).

krzentner commented:
This sounds fairly complex to achieve, since it will require all of those multi-task algorithms to somehow change the shape of their policies, even though the environments environment spec doesn't match. Currently only PEARL does that, and it's one of the things that make it more difficult to use.

I think we should drop this issue from the release, since it seems unlikely we can achieve this in the time remaining.
@ryanjulian What do you think?
